README.md
## Build and Deploy Spring Petclinic Application to AWS App Runner using AWS CDK.

## Overview
Leverage AWS services AWS App Runner, Amazon RDS, Amazon Elastic Container Registry, AWS CodePipeline, AWS CodeCommit, AWS CodeBuild, AWS CodeDeploy and AWS CDK to deploy infrastructure as code and build out a CI/CD pipeline. We will use Spring Boot Petclinic application as a sample app to demonstrate the CI/CD pipeline. Petclinic is an MVC application built using Spring framework and leverages a relational database to persist the data.

## Architecture
![Architecture](assets/architecture.png)

## Repository Structure
- petclinic - Spring Boot application code. This is a clone of the spring petclinic repo. Modification includes addition of the Dockerfile. For more details review the [README](petclinic/readme.md)
- cdk - AWS CDK Project for Building and Deploying Petclinic App. For more details review the [README](cdk/README.md)

## Configure Development Environment
- Follow the steps in the Cloud9 Environment Setup guide - [Create-Cloud9-Environment](CLOUD9-SETUP.md)

## Environment Setup and Local Testing

1. Setup and Install AWS CDK, Typescript:
    ```bash
    npm install -g typescript
    export CDK_DEFAULT_ACCOUNT="AWS ACCOUNT TO DEPLOY"
    export CDK_DEFAULT_REGION="AWS REGION TO DEPLOY"
    ```

2. Bootstrap AWS CDK:
    ```bash
    cdk bootstrap aws://$CDK_DEFAULT_ACCOUNT/$CDK_DEFAULT_REGION
    ```

3. Clone the source code repository from aws-samples:
    ```bash
    cd ~/environment
    git clone https://github.com/aws-samples/aws-apprunner-cdk.git
    ```

4. Setup Git User:
    ```
    git config --global user.name "workshop-admin"
    git config --global user.email "workshop-admin@example.com"
    ```

5. Init AWS CDK environment:
    ```
    cd ~/environment/aws-apprunner-cdk/cdk
    npm install @types/node
    ```

6. Package the application using Apache Maven Wrapper
    ```bash
    cd ~/environment/aws-apprunner-cdk/petclinic
    ./mvnw package -Dmaven.test.skip=true
    ```

7. Run Petclinic application locally and run the following inside the Cloud9 terminal:
    ```bash
    ./mvnw spring-boot:run
    ```

8. Validate the launched application and preview the application running locally:
    ![Petclinic Launch](assets/10-petclinic-preview-launch.png)

    This will run the application on port 8080. Click Preview from the top menu and then click “Preview Running Application.” It will open a browser displaying the Spring Petclinic application.

    ![Petclinic Preview](assets/11-petclinic-preview.png)

    Press Ctrl+C to stop the running application.


## Build the infrastructure and pipeline using AWS CDK and validate the application and pipeline.

**Note:** This workshop will create chargeable resources in your account. When finished, please make sure you clean up resources as instructed at the end.

1. Set up DockerHub Credentials in Secrets Manager for Docker Login credentials. Please provide your personal docker username (not email) and password. Note this step demonstrates the use of Docker Hub and provides an example on how to leverage the private registries to be included in the pipeline using Secrets Manager. 
    ```bash
    cd ~/environment/aws-apprunner-cdk
    aws secretsmanager create-secret \
        --name dockerhub_credentials \
        --description "DockerHub Credentials" \
        --secret-string "{\"username\":\"dockerhub_username\",\"password\":\"dockerhub_password\"}" \
        --region=$CDK_DEFAULT_REGION
    ```

2. Deploy Step Phase 1: Create Code Commit Repo
    - Setup the Code Commit Repo. 
        ```bash
        cd ~/environment/aws-apprunner-cdk/cdk
        cdk deploy
        ```
    - When asked whether you want to proceed with the actions, enter `y`.
        ![CDK Deploy](assets/12-c9-cdk-deploy.png)
    - Wait for AWS CDK to complete the deployment before proceeding. It will take few minutes to complete “cdk deploy”. Pipeline will fail since the code is not available in code commit repo.

3. Deploy Step Phase 2: Enable CI/CD Pipeline and deploy infrastructure and application.
    - Commit the code and set the origin to the code commit repo that was created.
        ```bash
        cd ~/environment/aws-apprunner-cdk
        git remote rename origin upstream
        git remote add origin "provide codecommit repo HTTPS URL created in above step"
        git push origin main
        ```
    - Code Commit changes will invoke the Code Pipeline. It will take approximately 20 to 30 minutes to complete the deployment.
    - Initial deployment will take longer since this includes building the container image for the application and setting up the infrastructure that includes Networking, RDS and other dependent components.
    - Explore the deployment progress on the CloudFormation console.

4. Validate the Deployment.
    - Once the deploy is complete, you can review the repo in App Runner service using the AWS console. 
    - View the App Runner service using the [AWS App Runner console](https://console.aws.amazon.com/apprunner/)

5. Review the deployed stack.
    - View the RDS database using the [Amazon RDS console](https://console.aws.amazon.com/rds).
    - View the ECR repo using the [Amazon ECR console](https://console.aws.amazon.com/ecr).
    - View the CodeCommit repo using the [AWS CodeCommit console](https://console.aws.amazon.com/codecommit).
    - View the CodeBuild project using the [AWS CodeBuild console](https://console.aws.amazon.com/codebuild).
    - View the CodePipeline using the [AWS CodePipeline console](https://console.aws.amazon.com/codepipeline).

6. Validate the CI/CD Pipeline.
    - Make a source code change in the petclinic application. Make the change to messages.properties under `petclinic/src/main/resources/messages/messages.properties`. Change the welcome value to `Welcome to CI/CD using AWS CDK`.
    - Commit the change and push the change to the code commit repo.
        ```bash
        cd ~/environment/aws-apprunner-cdk
        git commit -am "updated the title"
        git push origin main
        ```
    - Review that the commit started the code pipeline build. It will take approximately 10 minutes to deploy the change which includes building a new image and deploying the latest version to AWS AppRunner.
    - Once the deploy is complete, you can refresh the application home page and you should see the change reflected on the home page.

7. Cleaning Up.
    - Delete the AWS CDK stack.
        ```
        cd ~/environment/aws-apprunner-cdk/cdk
        cdk destroy
        ```
    - When asked whether you want to proceed with the actions, enter `y`. Wait for AWS CDK to complete the destroy.
    - Delete the AWS CloudFormation `petclinic-deploy-*` stack from the AWS CloudFormation Console.
    - Delete the images in the AWS ECR `petclinic` repo.
    - Delete the AWS CloudFormation `petclinic-build-*` stack from the AWS CloudFormation Console.
    - Delete the AWS Cloud9 `workshop-environment` environment including the `workshop-admin` IAM role.
    - Delete the `dockerhub_credentials` from AWS Secrets Manager.

## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This library is licensed under the MIT-0 License. See the LICENSE file.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-07-05

### Added

- Initial Release

## [1.0.1] - 2022-09-22

- Upgrade CDK version to 2.43.1.

## [1.0.2] - 2023-07-25

- Upgrade CDK version to 2.88.0 and dependency fixes.
- Deprecated code changes.

cdk/test/cdk-ecr-stack.test.ts
import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as CdkECRStage from "../lib/cdk-ecr-stack";

test("CDK ECR Stage", () => {
    const stackName = 'ecr-app-stage'
    const app = new cdk.App();

    // WHEN
    const stack = new CdkECRStage.CdkECRStack(app, stackName);

    // THEN
    const template = Template.fromStack(stack);

    // ASSERT
    template.resourceCountIs("AWS::ECR::Repository", 1);
    template.hasResourceProperties("AWS::ECR::Repository", {
        ImageScanningConfiguration: {
            "ScanOnPush": true
        },
        LifecyclePolicy: {
            "LifecyclePolicyText": "{\"rules\":[{\"rulePriority\":1,\"selection\":{\"tagStatus\":\"untagged\",\"countType\":\"sinceImagePushed\",\"countNumber\":1,\"countUnit\":\"days\"},\"action\":{\"type\":\"expire\"}},{\"rulePriority\":2,\"selection\":{\"tagStatus\":\"any\",\"countType\":\"imageCountMoreThan\",\"countNumber\":10},\"action\":{\"type\":\"expire\"}}]}"
        }
    });

});


cdk/test/cdk-infra-stack.test.ts
import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as CdkInfraStage from "../lib/cdk-infra-stack";

test("CDK Infra Stage", () => {
    const stackName = 'infra-stage'
    const app = new cdk.App();

    // WHEN
    const stack = new CdkInfraStage.CdkInfraStack(app, stackName);

    // THEN
    const template = Template.fromStack(stack);

    // Assert VPC and Subnets
    template.resourceCountIs("AWS::EC2::VPC", 1);
    template.hasResourceProperties("AWS::EC2::VPC", {
        CidrBlock: "10.0.0.0/26",
        EnableDnsHostnames: true,
        EnableDnsSupport: true,
    });
    template.resourceCountIs("AWS::EC2::Subnet", 2);
    template.hasResourceProperties("AWS::EC2::Subnet", {
        CidrBlock: "10.0.0.0/28",
        MapPublicIpOnLaunch: false,
    });

    // Assert DBInstance and Properties
    template.resourceCountIs("AWS::RDS::DBInstance", 1);
    template.hasResourceProperties("AWS::RDS::DBInstance", {
        DBInstanceClass: "db.t3.small",
        AllocatedStorage: "100",
        AllowMajorVersionUpgrade: false,
        AutoMinorVersionUpgrade: false,
        BackupRetentionPeriod: 5,
        Engine: "mysql",
        EngineVersion: "5.7",
        MultiAZ: true,
        PubliclyAccessible: false,
        StorageEncrypted: true,
        StorageType: "gp2",
    });

    // Assert DB Security Group
    template.resourceCountIs("AWS::EC2::SecurityGroup", 1);
    template.hasResourceProperties("AWS::EC2::SecurityGroup", {
        SecurityGroupEgress: [{
            CidrIp: "0.0.0.0/0",
        }],
        SecurityGroupIngress: [{
            FromPort: 3306,
            IpProtocol: "tcp",
            ToPort: 3306
        }]
    });

    // Assert AppRunner Service and VpcConnector
    template.resourceCountIs("AWS::AppRunner::Service", 1);
    template.hasResourceProperties("AWS::AppRunner::Service", {
        HealthCheckConfiguration: {
            Path: "/actuator/health"
        }
    });
    template.resourceCountIs("AWS::AppRunner::VpcConnector", 1);

});


cdk/test/cdk-pipeline-stack.test.ts
import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as CdkPipeline from "../lib/cdk-pipeline-stack";

test("CDK Pipeline Stack", () => {
    const stackName = 'petclinic'
    const app = new cdk.App();

    // WHEN
    const stack = new CdkPipeline.CdkPipelineStack(app, stackName);

    // THEN
    const template = Template.fromStack(stack);

    // Assert Code Commit Repo
    template.resourceCountIs("AWS::CodeCommit::Repository", 1);
    template.hasResourceProperties("AWS::CodeCommit::Repository", {
        RepositoryName: stackName,
    });

    // Assert Code Pipeline
    template.resourceCountIs("AWS::CodePipeline::Pipeline", 1);
    template.hasResourceProperties("AWS::CodePipeline::Pipeline", {
        Name: `${stackName}-pipeline`,
    });

    // Assert Code Build Project
    template.resourceCountIs("AWS::CodeBuild::Project", 2);
    template.hasResourceProperties("AWS::CodeBuild::Project", {
        "Environment": {
            "ComputeType": "BUILD_GENERAL1_MEDIUM",
            "PrivilegedMode": false,
            "Type": "LINUX_CONTAINER"
        },
    });

});


cdk/lib/cdk-app.ts
import { Stack } from "aws-cdk-lib";
import { BuildEnvironmentVariableType } from "aws-cdk-lib/aws-codebuild";

export class CdkApp {
  private stack: Stack;
  constructor(stack: Stack) {
    this.stack = stack;
  }

  /************************************************************************/
  /******************** prepare buildspec.yaml ****************************/
  /************************************************************************/
  public getBuildSpec() {
    return {
      version: "0.2",
      env: {
        variables: {
          IMAGE_TAG: "latest",
        },
        "exported-variables": [
          "IMAGE_TAG",
          "CODEBUILD_BUILD_NUMBER",
          "CODEBUILD_RESOLVED_SOURCE_VERSION",
        ],
      },
      phases: {
        install: {
          "runtime-versions": {
            java: "corretto8",
          },
          commands: [
            "echo running install commands...",
            "COMMIT_HASH=$(echo $CODEBUILD_RESOLVED_SOURCE_VERSION | cut -c 1-8)",
            "IMAGE_TAG=${COMMIT_HASH:=latest}",
          ],
        },
        pre_build: {
          commands: [
            "echo Logging in to Amazon ECR...",
            "echo $DOCKER_USER_PASSWORD | docker login -u $DOCKER_USER_NAME --password-stdin",
            "aws ecr get-login-password --region $AWS_DEFAULT_REGION | docker login --username AWS --password-stdin $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com",
          ],
        },
        build: {
          commands: [
            "echo Build started on `date`",
            "cd petclinic",
            "./mvnw package -Dmaven.test.skip=true",
            "docker build -t $IMAGE_REPO_NAME:$IMAGE_TAG .",
            "docker tag $IMAGE_REPO_NAME:$IMAGE_TAG $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:latest",
            "docker tag $IMAGE_REPO_NAME:$IMAGE_TAG $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG",
          ],
        },
        post_build: {
          commands: [
            "echo Running post build steps...",
            "docker push $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:latest",
            "docker push $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG",
          ],
        },
      },
      cache: {
        paths: "/root/.m2/**/*",
      },
    };
  }

  /************************************************************************/
  /******************** prepare build environment *************************/
  /************************************************************************/
  public getBuildEnvironment() {
    return {
      AWS_ACCOUNT_ID: {
        value: this.stack.account,
      },
      AWS_DEFAULT_REGION: {
        value: this.stack.region,
      },
      IMAGE_REPO_NAME: {
        value: this.stack.stackName,
      },
      IMAGE_TAG: {
        value: "latest",
      },
      DOCKER_USER_NAME: {
        type: BuildEnvironmentVariableType.SECRETS_MANAGER,
        value: "dockerhub_credentials:username",
      },
      DOCKER_USER_PASSWORD: {
        type: BuildEnvironmentVariableType.SECRETS_MANAGER,
        value: "dockerhub_credentials:password",
      },
    };
  }
}


cdk/lib/cdk-ecr-stack.ts
import { Construct } from "constructs";
import * as cdk from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import { TagStatus } from "aws-cdk-lib/aws-ecr";
import { Duration, Stack, StackProps } from "aws-cdk-lib";

export interface CdkECRStackProps extends StackProps{
  readonly appName?:string
}

export class CdkECRStack extends Stack {
  private appName:string
  constructor(scope: Construct, id: string, props?: CdkECRStackProps) {
    super(scope, id, props);
    if(props && props.appName){
      this.appName = props.appName
    }else{
      const appNameCtx = this.node.tryGetContext("app-name");
      //Generates Random ID in case context doesn't have app-name
      this.appName = appNameCtx ? appNameCtx : "ecr-app-name" + (Math.random() + 1).toString(36).substring(7);;
    }


    /************************************************************************/
    /************************* Create ECR Repo ******************************/
    /************************************************************************/
    const repository = new ecr.Repository(this, `${this.stackName}-build`, {
      repositoryName: this.appName,
      imageScanOnPush: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    /************************************************************************/
    /************************* Add Lifecycle Rules **************************/
    /************************************************************************/
    repository.addLifecycleRule({
      tagStatus: TagStatus.ANY,
      maxImageCount: 10
    });
    repository.addLifecycleRule({
      tagStatus: TagStatus.UNTAGGED,
      maxImageAge: Duration.days(1),
    });
  }
}


cdk/lib/cdk-pipeline-stack.ts
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { pipelines } from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as codecommit from "aws-cdk-lib/aws-codecommit";
import * as codebuild from "aws-cdk-lib/aws-codebuild";
import {
  CodeBuildStep,
  CodePipeline,
  CodePipelineSource,
} from "aws-cdk-lib/pipelines";
import { CdkApp } from "./cdk-app";
import { CdkPipelineECRStage } from "./cdk-pipeline-ecr-stage";
import { CdkPipelineInfraStage } from "./cdk-pipeline-infra-stage";

export class CdkPipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    /************************************************************************/
    /************************* Code Commit Repo *****************************/
    /************************************************************************/
    const repo = new codecommit.Repository(this, `${this.stackName}-repo`, {
      repositoryName: this.stackName,
    });

    /************************************************************************/
    /************************* Code Pipeline  *******************************/
    /************************************************************************/
    const pipeline = new CodePipeline(this, `${this.stackName}-pipeline`, {
      pipelineName: `${this.stackName}-pipeline`,
      selfMutation: false,
      crossAccountKeys: true,
      synth: new CodeBuildStep(`${this.stackName}-synth`, {
        projectName: `${this.stackName}-synth`,
        input: CodePipelineSource.codeCommit(repo, "main"),
        installCommands: ["npm install -g aws-cdk"],
        commands: ["cd cdk", "npm ci", "npm run build", "npx cdk synth"],
        primaryOutputDirectory: "cdk/cdk.out",
        buildEnvironment: {
          computeType: codebuild.ComputeType.MEDIUM,
        },
      }),
    });

    /************************************************************************/
    /************************* Code Build Role  *****************************/
    /************************************************************************/
    const buildRole = new iam.Role(this, `${this.stackName}-build-role`, {
      assumedBy: new iam.ServicePrincipal("codebuild.amazonaws.com"),
      description: `${this.stackName}-build-role`,
      inlinePolicies: {
        "petclinic-code-build-policy": new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "secretsmanager:GetSecretValue"
              ],
              resources: [
                `arn:${this.partition}:secretsmanager:${this.region}:${this.account}:secret:dockerhub_credentials`
              ],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "ecr:GetAuthorizationToken",
              ],
              resources: ["*"]
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:GetRepositoryPolicy",
                "ecr:DescribeRepositories",
                "ecr:ListImages",
                "ecr:DescribeImages",
                "ecr:BatchGetImage",
                "ecr:GetLifecyclePolicy",
                "ecr:GetLifecyclePolicyPreview",
                "ecr:ListTagsForResource",
                "ecr:DescribeImageScanFindings",
                "ecr:InitiateLayerUpload",
                "ecr:UploadLayerPart",
                "ecr:CompleteLayerUpload",
                "ecr:PutImage"
              ],
              resources: [
                `arn:${this.partition}:ecr:${this.region}:${this.account}:repository/petclinic`
              ],
            }),
          ]
        })
      }
    });

    /************************************************************************/
    /************************* Code Build Stage  ****************************/
    /************************************************************************/
    const app = new CdkApp(this);
    const build = new CdkPipelineECRStage(this, `${this.stackName}-build`);
    pipeline.addStage(build, {
      post: [
        new pipelines.CodeBuildStep(`${this.stackName}-build`, {
          projectName: `${this.stackName}-build`,
          partialBuildSpec: codebuild.BuildSpec.fromObject(app.getBuildSpec()),
          commands: ["echo building app"],
          buildEnvironment: {
            buildImage: codebuild.LinuxBuildImage.AMAZON_LINUX_2_3,
            computeType: codebuild.ComputeType.MEDIUM,
            privileged: true,
            environmentVariables: app.getBuildEnvironment(),
          },
          role: buildRole,
        }),
      ],
    });

    /************************************************************************/
    /************************* Code Deploy Stage  ***************************/
    /************************************************************************/
    const deploy = new CdkPipelineInfraStage(this, `${this.stackName}-deploy`);
    pipeline.addStage(deploy);
  }
}


cdk/lib/cdk-infra-stack.ts
import { Construct } from "constructs";
import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as rds from "aws-cdk-lib/aws-rds";
import * as apprunner from "aws-cdk-lib/aws-apprunner";
import { Stack, StackProps, Duration } from "aws-cdk-lib";
import * as secrets from "aws-cdk-lib/aws-secretsmanager";

export interface CdkInfraStackProps extends StackProps{
  readonly appName?:string;
}

export class CdkInfraStack extends Stack {
  private readonly appName:string;
  constructor(scope: Construct, id: string, props?: CdkInfraStackProps) {
    super(scope, id, props);
    if(props && props.appName){
      this.appName = props.appName;
    }else{
      const appNameCtx = this.node.tryGetContext("app-name");
      //Generates Random ID in case context doesn't have app-name
      this.appName =  appNameCtx ? appNameCtx : "infra-app-name" + (Math.random() + 1).toString(36).substring(7);
    }

    /************************************************************************/
    /****************************** VPC  ************************************/
    /************************************************************************/
    const vpc = new ec2.Vpc(this, `${this.stackName}-vpc`, {
      ipAddresses: ec2.IpAddresses.cidr('10.0.0.0/26'),
      maxAzs: 2,
      subnetConfiguration: [
        {
          cidrMask: 28,
          name: "private",
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    /************************************************************************/
    /***************************** MYSQLDB SG *******************************/
    /************************************************************************/
    const dbServerSG = new ec2.SecurityGroup(this, `${this.stackName}-rds-sg`, {
      vpc,
      allowAllOutbound: true,
      description: "Ingress for MySQL Server",
    });
    dbServerSG.addIngressRule(
      ec2.Peer.ipv4(vpc.vpcCidrBlock),
      ec2.Port.tcp(3306)
    );

    /************************************************************************/
    /****************************** MYSQL DB ********************************/
    /************************************************************************/
    const mysql = new rds.DatabaseInstance(
      this,
      `${this.stackName}-mysql-rds`,
      {
        engine: rds.DatabaseInstanceEngine.mysql({
          version: rds.MysqlEngineVersion.VER_5_7,
        }),
        instanceType: ec2.InstanceType.of(
          ec2.InstanceClass.BURSTABLE3,
          ec2.InstanceSize.SMALL
        ),
        credentials: rds.Credentials.fromGeneratedSecret(this.appName, {
          secretName: `rds/dev/${this.appName}/mysql`,
        }),
        vpc,
        vpcSubnets: {
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
        autoMinorVersionUpgrade: false,
        allowMajorVersionUpgrade: false,
        securityGroups: [dbServerSG],
        multiAz: true,
        backupRetention: Duration.days(5),
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        storageEncrypted: true,
      }
    );

    /************************************************************************/
    /************************ APPRUNNER Role and Service ********************/
    /************************************************************************/
    const appRunnerRole = new iam.Role(
      this,
      `${this.stackName}-apprunner-role`,
      {
        assumedBy: new iam.ServicePrincipal("build.apprunner.amazonaws.com"),
        description: `${this.stackName}-apprunner-role`,
        inlinePolicies: {
          "petclinic-apprunner-policy": new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: ["ecr:GetAuthorizationToken"],
                resources: ["*"],
              }),
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  "ecr:BatchCheckLayerAvailability",
                  "ecr:GetDownloadUrlForLayer",
                  "ecr:GetRepositoryPolicy",
                  "ecr:DescribeRepositories",
                  "ecr:ListImages",
                  "ecr:DescribeImages",
                  "ecr:BatchGetImage",
                  "ecr:GetLifecyclePolicy",
                  "ecr:GetLifecyclePolicyPreview",
                  "ecr:ListTagsForResource",
                  "ecr:DescribeImageScanFindings",
                ],
                resources: [
                  "arn:aws:ecr:" +
                    this.region +
                    ":" +
                    this.account +
                    ":repository/petclinic",
                ],
              }),
            ],
          }),
        },
      }
    );

    /************************************************************************/
    /************************ APPRUNNER VPCConnector ************************/
    /************************************************************************/
    const vpcConnector = new apprunner.CfnVpcConnector(this, "vpcConnector", {
      subnets: vpc.selectSubnets({
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      }).subnetIds,
      securityGroups: [dbServerSG.securityGroupId]
    });

    /************************************************************************/
    /************************ APPRUNNER Service *****************************/
    /************************************************************************/
    const appRunnerService = new apprunner.CfnService(
      this,
      `${this.stackName}-apprunner-service`,
      {
        serviceName: this.appName,
        sourceConfiguration: {
          authenticationConfiguration: {
            accessRoleArn: appRunnerRole.roleArn,
          },
          autoDeploymentsEnabled: true,
          imageRepository: {
            imageIdentifier: `${this.account}.dkr.ecr.${this.region}.amazonaws.com/${this.appName}:latest`,
            imageRepositoryType: "ECR",
            imageConfiguration: {
              port: "80",
              runtimeEnvironmentVariables: [
                {
                  name: "MYSQL_USER",
                  value: this.appName,
                },
                {
                  name: "MYSQL_PASS",
                  value: mysql.secret
                    ? mysql.secret.secretValueFromJson("password").toString()
                    : "password",
                },
                {
                  name: "MYSQL_URL",
                  value: `jdbc:mysql://${mysql.instanceEndpoint.hostname}/${this.appName}?createDatabaseIfNotExist=true`,
                },
              ],
            },
          },
        },
        instanceConfiguration: {
          cpu: "2048",
          memory: "4096",
        },
        healthCheckConfiguration: {
          path: "/actuator/health",
        },
        networkConfiguration: {
          egressConfiguration: {
            egressType: "VPC",
            vpcConnectorArn: vpcConnector.attrVpcConnectorArn,
          },
        },
      }
    );

    /************************************************************************/
    /************************** APPRUNNER Service URL ***********************/
    /************************************************************************/
    new cdk.CfnOutput(this, "serviceUrl", {
      value: appRunnerService.attrServiceUrl,
      exportName: "serviceUrl",
    });
  }
}
